powershell/VstsTaskSdk/ServerOMFunctions.ps1

```diff
@@ -48,10 +48,145 @@ $projectHttpClient.GetProjects().Result
+
+        [string]$Uri,
+
+        $VssCredentials)
+
+    Trace-EnteringInvocation -InvocationInfo $MyInvocation
+    try {
+        if (!$Uri) {
+            # Default the URI.
+            $Uri = Get-TaskVariable -Name System.TeamFoundationCollectionUri -Require
+        }
+
+        [uri]$Uri = New-Object System.Uri($Uri)
+
+        if (!$VssCredentials) {
+            # Default the credentials.
+            $VssCredentials = Get-VssCredentials
+        }
+
+        # Try to load the type.
+        try {
+            $null = [type]$TypeName
+        } catch {
+            Write-Verbose "Caught exception while attempting to load type: '$TypeName'"
+            if ($TypeName -like 'Microsoft.*.WebApi.*HttpClient') {
+                # Try to interpret the DLL name. Trim the last ".___HttpClient" and add ".dll".
+                $dll = $TypeName.SubString(0, $TypeName.LastIndexOf('.'))
+                $dll = [System.IO.Path]::Combine($OMDirectory, "$dll.dll")
+                Write-Verbose "Testing file path for interpreted assembly name: '$dll'"
+                if (!(Test-Path -LiteralPath $dll -PathType Leaf)) {
+                    Write-Verbose 'Not found. Rethrowing exception.'
+                    # Unable to find a file matching the interpreted assembly name.
+                    # Rethrow.
+                    throw
+                }
+
+                # Load the interpreted WebApi DLL.
+                Write-Verbose "Loading assembly: '$dll'"
+                Add-Type -LiteralPath $dll
+
+                # Try again to load the type, now that the WebApi DLL is loaded.
+                Write-Verbose "Re-attempting to load the type: '$TypeName'"
+                $null = [type]$TypeName
+            } else {
+                # Rethrow.
+                throw
+            }
+        }
+
+        # Try to construct the HTTP client.
+        try {
+            New-Object $TypeName($Uri, $VssCredentials)
+        } catch {
+            # Check if the exception is due to Newtonsoft.Json DLL not found.
+            if ($_.Exception.InnerException -isnot [System.IO.FileNotFoundException] -or
+                $_.Exception.InnerException.FileName -notlike 'Newtonsoft.Json, *') {
+
+                # Rethrow.
+                throw
+            }
+
+            # Test if the Newtonsoft.Json DLL exists in the OM directory.
+            $newtonsoftDll = [System.IO.Path]::Combine($OMDirectory, "Newtonsoft.Json.dll")
+            Write-Verbose "Testing file path: '$newtonsoftDll'"
+            if (!(Test-Path -LiteralPath $newtonsoftDll -PathType Leaf)) {
+                Write-Verbose 'Not found. Rethrowing exception.'
+                throw
+            }
+
+            # Add a binding redirect and try again. Parts of the Dev15 preview SDK have a
+            # dependency on the 6.0.0.0 Newtonsoft.Json DLL, while other parts reference
+            # the 8.0.0.0 Newtonsoft.Json DLL.
+            Write-Verbose "Adding assembly resolver."
```

Comment:
I would add the assembly resolver logic unconditionally, just like the VsoWorker does through its vsoworker.exe.config. There's also a redirect for System.Net.Http.Formatting which we haven't crashed on, but it may be worth adding.